script/ensure-node-source-fixture-files.js

```diff
@@ -0,0 +1,163 @@
+    filename: "SHASUMS256.txt",
+    sha256sum: "64aad1211a6003dd6529ebf9f19167769d0356ce5affc4245bb26c35aa66a9ed"
+  },
+]; // If you ever need to update to a newer Node version, just calculate the sha256sums
+   // for the new files and update them here to match the new files.
+
+const sourceFixtureDir = path.resolve(__dirname, "..", "spec", "fixtures", "node-source");
+fs.mkdirSync(sourceFixtureDir, { recursive: true });
+
+for (const details of filesToFetch) {
+  ensureFile(details);
+} // This is the main loop, with support functions below.
+
+async function ensureFile(details) {
+  // We check if the file already exists, and if so,
+  // make sure its sha256sum matches the expected/correct value.
+  // If it doesn't exist or has the wrong hash, try to re-download.
+
+  logVerbose("details is:");
+  logVerbose(details);
+
+  const destinationPath = path.resolve(sourceFixtureDir, details.filename);
+  logVerbose(`destinationPath is: ${destinationPath}`);
+
+  const existingFileIsCorrect = await verifyExistingFile(destinationPath, details.sha256sum);
+
+  if (!existingFileIsCorrect) {
+    logVerbose("Hash did not match, re-downloading...");
+    // Get the file
+    downloadFileToDestination(details.url, destinationPath)
+      .then(async function () {
+        console.log(`Successfully downloaded file from ${details.url}.`);
+        logVerbose("checking if hash matches in for...of loop.");
+        // Check its hash
+        const hashDidMatch = await verifyHash(destinationPath, details.sha256sum);
+        if (!hashDidMatch) {
+          console.error(`Hash did not match for ${destinationPath}`);
+        }
+      })
+      .catch(console.error);
+  }
+}
+
+async function verifyExistingFile(targetPath, expectedHash) {
+  if (fs.existsSync(targetPath)) {
+    logVerbose(`${targetPath} already exists.`);
+
+    logVerbose(`verifying hash for ${targetPath} in verifyExistingFile`);
+    const hashDidMatch = await verifyHash(targetPath, expectedHash);
+
+    if (hashDidMatch) {
+      // Successfully verified existing file. Return true.
+      console.log(`Existing file ${targetPath} successfully verified.`);
+      return true;
+    } else {
+      // Existing file's hash was wrong. Delete the old file.
+      console.error(`Hash did not match for ${targetPath}. Deleting.`);
+      fs.rmSync(targetPath);
+    }
+  } else {
+    // File did not actually exist.
+    logVerbose(`${targetPath} does not exist.`);
+  }
+  // If we haven't returned yet, verification did not succeed. Return false.
+  return false;
+}
+
+async function verifyHash(path, expectedHash) {
+  // Checks the hash of a given locally downloaded file against its expected value.
+  // Returns true/false.
+  // See: https://nodejs.org/api/crypto.html#class-hash for details.
+
+  // This module apparently needs to be required each time this function runs,
+  // or hashing multiple files in parallel can get messed up? Not sure why.
+  const { createHash } = require("node:crypto");
```

Comment:
I'm glad you left the comment here, otherwise I was about to write up worried about unneeded resource usage.

But I've seen that before funnily enough in native modules that aren't context aware, or what's more likely is this builder works with a globally defined object (in relation to the actual module) and since NodeJS will require the same module it would likely actually retain it's state, including any global variables declared. So probably why it gets all worried about being used multiple times simultaneously.

Reply:
Yeah, I'm convinced there's a proper way to clear that reused state, but I tried a bunch of things and wasn't able to figure it out. I would not at all mind a fix for this that only `require()`s it once. But I couldn't figure it out myself, personally. All the hashes were coming out wrong starting at the second file I tried to hash.